src/gtc/passes/gtir_definitive_assignment_analysis.py

```diff
@@ -0,0 +1,63 @@
+import copy
+import warnings
+from typing import List, Set
+
+from eve import NodeVisitor
+from gtc import gtir
+
+
+class DefinitiveAssignmentAnalysis(NodeVisitor):
+    def visit_IfStmt(self, node: gtir.FieldIfStmt, *, alive_vars: Set[str], **kwargs):
+        true_branch_vars = copy.copy(alive_vars)
+        false_branch_vars = copy.copy(alive_vars)
+        self.visit(node.true_branch, alive_vars=true_branch_vars, **kwargs)
+        self.visit(node.false_branch, alive_vars=false_branch_vars, **kwargs)
+        alive_vars.update(true_branch_vars & false_branch_vars)
+
+    def visit_ParAssignStmt(self, node: gtir.ParAssignStmt, *, alive_vars: Set[str], **kwargs):
+        self.visit(node.right, alive_vars=alive_vars, **kwargs)
+        alive_vars.add(node.left.name)
+
+    def visit_FieldAccess(
+        self,
+        node: gtir.FieldAccess,
+        *,
+        alive_vars: Set,
+        invalid_accesses: List[gtir.FieldAccess],
+        **kwargs,
+    ):
+        if node.name not in alive_vars:
+            invalid_accesses.append(node)
+
+
+def analyse(gtir_stencil_expr: gtir.Stencil):
```

Comment:
It's not really documented but the (implicit) convention is to declare an `apply()` class method in the visitor as the public interface (creating an instance and executing it with the provided data). If a free function API is also desired, it's just an alias to apply, like:

```python
analyse = DefinitiveAssignmentAnalysis.apply
```
On a different topic, I think we use US spelling everywhere in the code, and in that case `analyze` fits better.